### PR TITLE
Ross: Time Entry Description Autocomplete Mode

### PR DESCRIPTION
**Pull request by masterrr**
Provides a user an autocomplete view when editing a time entry description.

All remarks from https://github.com/toggl/mobile/pull/363 are met.
Resolves https://github.com/toggl/mobile/issues/110

### CODE REVIEW

File: Ross/Views/SuggestionTableViewCell.cs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                }
+            } else {
+                Console.WriteLine ("project model is null");
+            }
+
+            projectLabel.TextColor = projectColor;
+
+            Console.WriteLine ("Checking {0}", projectName);
+            if (projectLabel.Text != projectName) {
+                Console.WriteLine ("Setting project label text {0}", projectName);
```
Comment (monday8am) [inline Ross/Views/SuggestionTableViewCell.cs]: Remove Console.WriteLine!

**Comment by monday8am** [comment]
- Cell text superposition ( check image below)
- If I select a cell, sometimes the rest of variables are set but not the description. I write test, select a cell with description "testing" and still the description  has the "test" value.
- Remove all Console.WriteLine
- Organise ALL imports


screen shot 2015-02-05 at 10 07 38

**Comment by masterrr** [comment]
@monday8am all done!

File: Ross/ViewControllers/LogViewController.cs [hunk 2]
```diff
@@ -174,6 +174,16 @@ private void OnContinue (TimeEntryModel model)
...
+
+            protected override void Update ()
+            {
+                CATransaction.Begin ();
+                CATransaction.CompletionBlock = delegate {
+                    TableView.ReloadData ();
+                };
+                base.Update ();
+                CATransaction.Commit();
+            }
```
Comment (monday8am) [inline Ross/ViewControllers/LogViewController.cs]: Why this block here?

File: Phoebe/Phoebe.iOS.csproj [hunk 3]
```diff
@@ -62,6 +62,7 @@
     <Compile Include="Threading\**\*.cs" />
     <Compile Include="Analytics\**\*.cs" />
     <Compile Include="Logging\**\*.cs" />
+    <Compile Include="Data\Views\SuggestionEntriesView.cs" />
```
Comment (monday8am) [inline Phoebe/Phoebe.iOS.csproj]: Please, remove this extra line :)

File: Phoebe/Data/Views/SuggestionEntriesView.cs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+        public bool HasSuggestions
+        {
+            get { return FilteredEntries != null && FilteredEntries.Count > 0; }
+        }
+
+        public void Dispose()
+        {
+
+        }
```
Comment (monday8am) [inline Phoebe/Data/Views/SuggestionEntriesView.cs]: Empty method. If is empty, just remove it.

File: Phoebe/Data/Views/SuggestionEntriesView.cs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+        public SuggestionEntriesView ()
+        {
+            Reload ();
+            IsLoading = false;
```
Comment (monday8am) [inline Phoebe/Data/Views/SuggestionEntriesView.cs]: I think "IsLoading = false;" is not needed. You have the same assignment in a finally inside Reload method.

File: Phoebe/Data/Views/SuggestionEntriesView.cs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        public void LoadMore ()
+        {
+            Reload ();
+        }
+
+        public void Reload()
+        {
+            if (IsLoading) {
+                return;
+            }
```
Comment (monday8am) [inline Phoebe/Data/Views/SuggestionEntriesView.cs]: the condition is duplicated. You have the same condition inside the method.

File: Ross/ViewControllers/SyncStatusViewController.cs [hunk 7]
```diff
@@ -118,6 +118,7 @@ public override void ViewDidLoad ()
             base.ViewDidLoad ();
 
             // Add content view controller:
+            contentViewController.WillMoveToParentViewController (this);
```
Comment (monday8am) [inline Ross/ViewControllers/SyncStatusViewController.cs]: Please, organise imports in this class.

File: Ross/ViewControllers/LogViewController.cs (hunk 2 above)
Comment (masterrr) [inline Ross/ViewControllers/LogViewController.cs]: To make sure all data is re-loaded properly. 

Underneath there is a method that manually adding / updating / deleting rows, after updating the tags it was one or two times a “tag problem” occured to me: the tags icon disappeared from where it should be. 

This approach waits until all insert / deleting animations are done and then reload the tableView w/o any animation (so if there is thing the underneath method doesn't do properly, it'll be onscreen after the harsh reload).

Comment (monday8am) [inline Ross/ViewControllers/LogViewController.cs]: :+1:

**Comment by roosmaa** [comment]
Is it just me or has the layout been broken in edit TE view?

ios simulator screen shot 09 feb 2015 13 05 06

Delete button used to be at the bottom of the screen. Or is this branch based on master which had broken layouts anyway? (maybe rebase if that is the case?)

File: Ross/ViewControllers/EditTimeEntryViewController.cs [hunk 8]
```diff
@@ -39,6 +59,8 @@ public class EditTimeEntryViewController : UIViewController
         private bool autoCommitScheduled;
         private int autoCommitId;
         private bool shouldRebindOnAppear;
+        private UITableView autoCompletionTableView;
+        private UIBarButtonItem autoCopmletionDoneBarButtonItem;
```
Comment (roosmaa) [inline Ross/ViewControllers/EditTimeEntryViewController.cs]: Typo in var name (Copmletion).

**Comment by masterrr** [comment]
@roosmaa what branch should be used as a parent? this one is on master, yes

File: Ross/ViewControllers/SyncStatusViewController.cs (hunk 7 above)
Comment (masterrr) [inline Ross/ViewControllers/SyncStatusViewController.cs]: Done.

File: Ross/ViewControllers/EditTimeEntryViewController.cs [hunk 9]
```diff
@@ -359,6 +503,12 @@ public override void LoadView ()
             deleteButton.SetTitle ("EditEntryDelete".Tr (), UIControlState.Normal);
             deleteButton.TouchUpInside += OnDeleteButtonTouchUpInside;
 
+            wrapper.Add (autoCompletionTableView = new TGTableView() {
+                TranslatesAutoresizingMaskIntoConstraints = false,
+                EstimatedRowHeight = 60.0f,
+                BackgroundColor = UIColor.Clear
+            } .Apply (BindAutocompletionTableView));
+
```
Comment (monday8am) [inline Ross/ViewControllers/EditTimeEntryViewController.cs]: @masterrr 
if you put the new autoCompletionTableView in last position then it breaks the previous layout ( with Delete button at bottom).

File: Ross/ViewControllers/LogViewController.cs (hunk 2 above)
Comment (monday8am) [inline Ross/ViewControllers/LogViewController.cs]: Remember to organise imports

**Comment by monday8am** [comment]
@masterrr  another detail, if you're editing time the autocompletion should be disabled or better, you should hide datePicker when the description text receive a focus event. 
Is looking great btw :)

File: Ross/ViewControllers/EditTimeEntryViewController.cs (hunk 9 above)
Comment (masterrr) [inline Ross/ViewControllers/EditTimeEntryViewController.cs]: Right. Tested & fixed.

**Comment by monday8am** [comment]
@masterrr please, organise imports in this class:

https://github.com/masterrr/mobile/blob/masterrr-autocomplete-2/Ross/ViewControllers/LogViewController.cs

**Comment by masterrr** [comment]
@monday8am sorted

**Comment by masterrr** [comment]
@monday8am now the description field dismisses the datetime picker, and vice versa

**Comment by monday8am** [comment]
@masterrr nice, great job :) For me the code is ok.

@taavirehemagi, please, check the code and test it.

**Comment by taavirehemagi** [comment]
For me it looks good! :+1:

**Pull request merged**